In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib
import seaborn as sns
import re

In [45]:
filename = 'partial_reviews1209.csv'
dataor = pd.read_csv(filename,encoding='ISO-8859-1')
dataor

Unnamed: 0  listing_id         id        date  reviewer_id  \
0              0        2515        198  2008-10-13         2603   
1              1        2515        859  2009-03-08         8455   
2              2        2515       1083  2009-03-25         9759   
3              3        2515       1107  2009-03-27         9193   
4              4        2515       2175  2009-05-09         7048   
5              5        2515       2889  2009-05-20        12521   
6              6        2515       3213  2009-05-27         5889   
7              7        2515       3313  2009-05-30        15342   
8              8        2515       7144  2009-08-10        26899   
9              9        2515       7350  2009-08-13        26718   
10            10        2515       8463  2009-08-29         6407   
11            11        2515      11177  2009-09-24        21062   
12            12        2515      11651  2009-09-27        39004   
13            13        2515      14145  2009-10-18        45184   
14            14        2515      14349  2009-10-20        35090   
15            15        2515      16302  2009-11-05        44089   
16            16        2515      25151  2010-01-31        68057   
17            17        2515      25500  2010-02-05      1470208   
18            18        2515      28524  2010-03-07        76069   
19            19        2515      28930  2010-03-10        80901   
20            20        2515      40150  2010-05-05        24048   
21            21        2515      46980  2010-05-28        71663   
22            22        2515      51242  2010-06-09       116704   
23            23        2515      59526  2010-07-01        70724   
24            24        2515      64294  2010-07-13       140772   
25            25        2515      66624  2010-07-19       138191   
26            26        2515      74208  2010-08-05       182441   
27            27        2515      77637  2010-08-13       183527   
28            28        2515      81511  2010-08-20        96782   
29            29        2515      95059  2010-09-12        68468   
...          ...         ...        ...         ...          ...   
9970        9970       23135  176797209  2017-07-31    125868928   
9971        9971       23135  187199739  2017-08-26      8694516   
9972        9972       23135  198765967  2017-09-30     40388828   
9973        9973       23135  200466871  2017-10-05    115674703   
9974        9974       23135  201947492  2017-10-09     24332672   
9975        9975       23135  204127940  2017-10-16    133386797   
9976        9976       23135  206483985  2017-10-25     89347821   
9977        9977       23135  208000131  2017-10-30     29067391   
9978        9978       23135  208704743  2017-11-02       345161   
9979        9979       23135  209906875  2017-11-06     19583260   
9980        9980       23135  214321166  2017-11-25     67255754   
9981        9981       23135  218891332  2017-12-15     63114422   
9982        9982       23135  221711447  2017-12-27     16574612   
9983        9983       23135  242607538  2018-03-12     45442987   
9984        9984       23135  245042307  2018-03-20      4105899   
9985        9985       23135  247717299  2018-03-29      7096411   
9986        9986       23135  251235341  2018-04-07     15632373   
9987        9987       23135  256777814  2018-04-23    114675472   
9988        9988       23135  259228108  2018-04-30    178962875   
9989        9989       23135  262624083  2018-05-09    118210804   
9990        9990       23135  264235669  2018-05-13       515013   
9991        9991       23135  266702241  2018-05-20     12795426   
9992        9992       23135  269580912  2018-05-27      3191318   
9993        9993       23135  277568127  2018-06-16     19811170   
9994        9994       23135  281287657  2018-06-24      5097599   
9995        9995       23135  292663736  2018-07-18    194270361   
9996        9996       23135  299288547  2018-07-30     165364

# Clean data

check data types

In [46]:
dataor.dtypes

Unnamed: 0        int64
listing_id        int64
id                int64
date             object
reviewer_id       int64
reviewer_name    object
comments         object
dtype: object

In [47]:
#change data type and rename column name
dataor["date"]=dataor["date"].astype('datetime64[ns]')
dataor.rename(columns={'Unnamed: 0':'original_No.'},inplace=True)
dataor.dtypes

original_No.              int64
listing_id                int64
id                        int64
date             datetime64[ns]
reviewer_id               int64
reviewer_name            object
comments                 object
dtype: object

In [49]:
#drop NULL 
dataor=dataor.dropna() 

In [50]:
for elem in dataor["comments"]:
    if len(elem)<2:
        print (elem)

!
N
P


In [53]:
#drop meaningless reviews like "!" "N" "P"
#reset index
dataor=dataor[dataor['comments']!='!']
dataor=dataor[dataor['comments']!='N']
dataor=dataor[dataor['comments']!='P']              
dataor.reset_index(drop=True,inplace=True)
dataor

original_No.  listing_id         id       date  reviewer_id  \
0                0        2515        198 2008-10-13         2603   
1                1        2515        859 2009-03-08         8455   
2                2        2515       1083 2009-03-25         9759   
3                3        2515       1107 2009-03-27         9193   
4                4        2515       2175 2009-05-09         7048   
5                5        2515       2889 2009-05-20        12521   
6                6        2515       3213 2009-05-27         5889   
7                7        2515       3313 2009-05-30        15342   
8                8        2515       7144 2009-08-10        26899   
9                9        2515       7350 2009-08-13        26718   
10              10        2515       8463 2009-08-29         6407   
11              11        2515      11177 2009-09-24        21062   
12              12        2515      11651 2009-09-27        39004   
13              13        2515      14145 2009-10-18        45184   
14              14        2515      14349 2009-10-20        35090   
15              15        2515      16302 2009-11-05        44089   
16              16        2515      25151 2010-01-31        68057   
17              17        2515      25500 2010-02-05      1470208   
18              18        2515      28524 2010-03-07        76069   
19              19        2515      28930 2010-03-10        80901   
20              20        2515      40150 2010-05-05        24048   
21              22        2515      51242 2010-06-09       116704   
22              23        2515      59526 2010-07-01        70724   
23              24        2515      64294 2010-07-13       140772   
24              25        2515      66624 2010-07-19       138191   
25              26        2515      74208 2010-08-05       182441   
26              27        2515      77637 2010-08-13       183527   
27              28        2515      81511 2010-08-20        96782   
28              29        2515      95059 2010-09-12        68468   
29              30        2515     100036 2010-09-19        69604   
...            ...         ...        ...        ...          ...   
9949          9970       23135  176797209 2017-07-31    125868928   
9950          9971       23135  187199739 2017-08-26      8694516   
9951          9972       23135  198765967 2017-09-30     40388828   
9952          9973       23135  200466871 2017-10-05    115674703   
9953          9974       23135  201947492 2017-10-09     24332672   
9954          9975       23135  204127940 2017-10-16    133386797   
9955          9976       23135  206483985 2017-10-25     89347821   
9956          9977       23135  208000131 2017-10-30     29067391   
9957          9978       23135  208704743 2017-11-02       345161   
9958          9979       23135  209906875 2017-11-06     19583260   
9959          9980       23135  214321166 2017-11-25     67255754   
9960          9981       23135  218891332 2017-12-15     63114422   
9961          9982       23135  221711447 2017-12-27     16574612   
9962          9983       23135  242607538 2018-03-12     45442987   
9963          9984       23135  245042307 2018-03-20      4105899   
9964          9985       23135  247717299 2018-03-29      7096411   
9965          9986       23135  251235341 2018-04-07     15632373   
9966          9987       23135  256777814 2018-04-23    114675472   
9967          9988       23135  259228108 2018-04-30    178962875   
9968          9989       23135  262624083 2018-05-09    118210804   
9969          9990       23135  264235669 2018-05-13       515013   
9970          9991       23135  266702241 2018-05-20     12795426   
9971          9992       23135  269580912 2018-05-27      3191318   
9972          9993       23135  277568127 2018-06-16     19811170   
9973          9994       23135  281287657 2018-06-24      5097599   
9974          9995       23135  292663736 2018-07-18    194270361   
9975

In [57]:
#removing numbers
#changing to lower case
def clean(string):
    string = re.sub(r"\d", "", string) # remove numbers  
    string = re.sub(r"_+", "", string) # remove consecutive underscores
    string = string.lower() # tranform to lower case    
    
    return string.strip()

dataor["cleancomments"] = dataor.comments.apply(clean)
dataor.head()

original_No.  listing_id    id       date  reviewer_id reviewer_name  \
0             0        2515   198 2008-10-13         2603         Jenny   
1             1        2515   859 2009-03-08         8455        Roland   
2             2        2515  1083 2009-03-25         9759           Cem   
3             3        2515  1107 2009-03-27         9193         Holly   
4             4        2515  2175 2009-05-09         7048    Alessandra   

                                            comments  \
0  Stephanie was a wonderful host! Her apartment ...   
1  Such a wonderful place and very close to the m...   
2  I just got back from a trip to NYC during whic...   
3  Stephanie's offered all the most important thi...   
4  Stephanie was really nice, ftiendly and helpfu...   

                                       cleancomments  
0  stephanie was a wonderful host! her apartment ...  
1  such a wonderful place and very close to the m...  
2  i just got back from a trip to nyc during whic...  
3  stephanie's offered all the most important thi...  
4  stephanie was really nice, ftiendly and helpfu...

# Detect language of revews and translate non-English reviews to English

In [58]:
# -*- coding: utf-8 -*-
import os, requests, uuid, json
subscriptionKey = '...' #USER KEY
base_url = 'https://api.cognitive.microsofttranslator.com'
path = '/translate?api-version=3.0'
params = '&to=en'
constructed_url = base_url + path + params
headers = {
    'Ocp-Apim-Subscription-Key': subscriptionKey,
    'Content-type': 'application/json',
    'X-ClientTraceId': str(uuid.uuid4())
}
#body = [{'text': ""}]
#request = requests.post(constructed_url, headers=headers, json=body)
#response = request.json()

select first 1000 reviews to see the percentage of non-English reviews

In [61]:
listco=dataor["cleancomments"][0:1000].tolist()
listnon=[]

for elem in range(len(listco)):
    listco[elem]=str(listco[elem])
    body = [{'text': listco[elem]}]
    request = requests.post(constructed_url, headers=headers, json=body)
    response = request.json()
    if response[0]['detectedLanguage']['language']!="en":
        listnon.append(listco[elem])
        #listc[elem]=translator.translate(listc[elem])
    #listc[elem]=response[0]['translations'][0]['text']

In [66]:
print(len(listnon)/len(listco)*100,"%")

16.900000000000002 %


for the first 1000 reviews, 16.9% of them are non-English reviews, so we cannot delete them directly, we need to translate them into English reviews

In [69]:
listclean=dataor["cleancomments"][0:2000].tolist()
for elem in range(len(listclean)):
    listclean[elem]=str(listclean[elem])
    body = [{'text': listclean[elem]}]
    request = requests.post(constructed_url, headers=headers, json=body)
    response = request.json()
    if response[0]['detectedLanguage']['language']!="en":
        listclean[elem]=response[0]['translations'][0]['text']

In [70]:
listclean

['stephanie was a wonderful host! her apartment was beautiful and extremely clean. we had a very nice (and affordable!) stay in the city with private room and own bathroom.',
 "Such a wonderful place and very close to the metro station on the line. It was a pleasure to stay in. She's French, so last word...\r\nThanks again Stephanie for your hospitality. We had a great time with you. Very ¨ s available for us with plenty of advice, I would especially recommend this address which is more for a reasonable price. The Metro is just a block from the apartment and the line--is the most central and convenient NYC. His neighborhood is really nice too!!! Go there closed eyes.",
 'i just got back from a trip to nyc during which i stayed at stephanie\'s appartment. the bedroom and private bathroom that comes with it were both kept very clean and are spacious so that i could unpack my stuff and put my clothes away in the closet. i didn\'t have to live out of my rucksack for ten days which made me 

In [71]:
listclean2=dataor["cleancomments"][2000:4000].tolist()
for elem in range(len(listclean2)):
    listclean2[elem]=str(listclean2[elem])
    body = [{'text': listclean2[elem]}]
    request = requests.post(constructed_url, headers=headers, json=body)
    response = request.json()
    if response[0]['detectedLanguage']['language']!="en":
        listclean2[elem]=response[0]['translations'][0]['text']

In [74]:
# -*- coding: utf-8 -*-
import os, requests, uuid, json
subscriptionKey = '...' #USER KEY
base_url = 'https://api.cognitive.microsofttranslator.com'
path = '/translate?api-version=3.0'
params = '&to=en'
constructed_url = base_url + path + params
headers = {
    'Ocp-Apim-Subscription-Key': subscriptionKey,
    'Content-type': 'application/json',
    'X-ClientTraceId': str(uuid.uuid4())
}

In [75]:
listclean3=dataor["cleancomments"][4000:].tolist()
for elem in range(len(listclean3)):
    listclean3[elem]=str(listclean3[elem])
    body = [{'text': listclean3[elem]}]
    request = requests.post(constructed_url, headers=headers, json=body)
    response = request.json()
    if response[0]['detectedLanguage']['language']!="en":
        listclean3[elem]=response[0]['translations'][0]['text']

In [86]:
listcleann=listclean+listclean2+listclean3

In [87]:
print(len(listcleann))

9979


In [88]:
dataor["transcomments"]=pd.DataFrame(listcleann)

In [89]:
dataor

original_No.  listing_id         id       date  reviewer_id  \
0                0        2515        198 2008-10-13         2603   
1                1        2515        859 2009-03-08         8455   
2                2        2515       1083 2009-03-25         9759   
3                3        2515       1107 2009-03-27         9193   
4                4        2515       2175 2009-05-09         7048   
5                5        2515       2889 2009-05-20        12521   
6                6        2515       3213 2009-05-27         5889   
7                7        2515       3313 2009-05-30        15342   
8                8        2515       7144 2009-08-10        26899   
9                9        2515       7350 2009-08-13        26718   
10              10        2515       8463 2009-08-29         6407   
11              11        2515      11177 2009-09-24        21062   
12              12        2515      11651 2009-09-27        39004   
13              13        2515      14145 2009-10-18        45184   
14              14        2515      14349 2009-10-20        35090   
15              15        2515      16302 2009-11-05        44089   
16              16        2515      25151 2010-01-31        68057   
17              17        2515      25500 2010-02-05      1470208   
18              18        2515      28524 2010-03-07        76069   
19              19        2515      28930 2010-03-10        80901   
20              20        2515      40150 2010-05-05        24048   
21              22        2515      51242 2010-06-09       116704   
22              23        2515      59526 2010-07-01        70724   
23              24        2515      64294 2010-07-13       140772   
24              25        2515      66624 2010-07-19       138191   
25              26        2515      74208 2010-08-05       182441   
26              27        2515      77637 2010-08-13       183527   
27              28        2515      81511 2010-08-20        96782   
28              29        2515      95059 2010-09-12        68468   
29              30        2515     100036 2010-09-19        69604   
...            ...         ...        ...        ...          ...   
9949          9970       23135  176797209 2017-07-31    125868928   
9950          9971       23135  187199739 2017-08-26      8694516   
9951          9972       23135  198765967 2017-09-30     40388828   
9952          9973       23135  200466871 2017-10-05    115674703   
9953          9974       23135  201947492 2017-10-09     24332672   
9954          9975       23135  204127940 2017-10-16    133386797   
9955          9976       23135  206483985 2017-10-25     89347821   
9956          9977       23135  208000131 2017-10-30     29067391   
9957          9978       23135  208704743 2017-11-02       345161   
9958          9979       23135  209906875 2017-11-06     19583260   
9959          9980       23135  214321166 2017-11-25     67255754   
9960          9981       23135  218891332 2017-12-15     63114422   
9961          9982       23135  221711447 2017-12-27     16574612   
9962          9983       23135  242607538 2018-03-12     45442987   
9963          9984       23135  245042307 2018-03-20      4105899   
9964          9985       23135  247717299 2018-03-29      7096411   
9965          9986       23135  251235341 2018-04-07     15632373   
9966          9987       23135  256777814 2018-04-23    114675472   
9967          9988       23135  259228108 2018-04-30    178962875   
9968          9989       23135  262624083 2018-05-09    118210804   
9969          9990       23135  264235669 2018-05-13       515013   
9970          9991       23135  266702241 2018-05-20     12795426   
9971          9992       23135  269580912 2018-05-27      3191318   
9972          9993       23135  277568127 2018-06-16     19811170   
9973          9994       23135  281287657 2018-06-24      5097599   
9974          9995       23135  292663736 2018-07-18    194270361   
9975

In [90]:
dataor.to_csv("partial_review_clean.csv")

# Measuring the success of translation

In [2]:
filename = 'review_clean.csv'
dataor1 = pd.read_csv(filename,encoding='ISO-8859-1')
dataor1

Unnamed: 0  original_No.  listing_id         id        date  \
0              0             0        2515        198  2008-10-13   
1              1             1        2515        859  2009-03-08   
2              2             2        2515       1083  2009-03-25   
3              3             3        2515       1107  2009-03-27   
4              4             4        2515       2175  2009-05-09   
5              5             5        2515       2889  2009-05-20   
6              6             6        2515       3213  2009-05-27   
7              7             7        2515       3313  2009-05-30   
8              8             8        2515       7144  2009-08-10   
9              9             9        2515       7350  2009-08-13   
10            10            10        2515       8463  2009-08-29   
11            11            11        2515      11177  2009-09-24   
12            12            12        2515      11651  2009-09-27   
13            13            13        2515      14145  2009-10-18   
14            14            14        2515      14349  2009-10-20   
15            15            15        2515      16302  2009-11-05   
16            16            16        2515      25151  2010-01-31   
17            17            17        2515      25500  2010-02-05   
18            18            18        2515      28524  2010-03-07   
19            19            19        2515      28930  2010-03-10   
20            20            20        2515      40150  2010-05-05   
21            21            22        2515      51242  2010-06-09   
22            22            23        2515      59526  2010-07-01   
23            23            24        2515      64294  2010-07-13   
24            24            25        2515      66624  2010-07-19   
25            25            26        2515      74208  2010-08-05   
26            26            27        2515      77637  2010-08-13   
27            27            28        2515      81511  2010-08-20   
28            28            29        2515      95059  2010-09-12   
29            29            30        2515     100036  2010-09-19   
...          ...           ...         ...        ...         ...   
9949        9949          9970       23135  176797209  2017-07-31   
9950        9950          9971       23135  187199739  2017-08-26   
9951        9951          9972       23135  198765967  2017-09-30   
9952        9952          9973       23135  200466871  2017-10-05   
9953        9953          9974       23135  201947492  2017-10-09   
9954        9954          9975       23135  204127940  2017-10-16   
9955        9955          9976       23135  206483985  2017-10-25   
9956        9956          9977       23135  208000131  2017-10-30   
9957        9957          9978       23135  208704743  2017-11-02   
9958        9958          9979       23135  209906875  2017-11-06   
9959        9959          9980       23135  214321166  2017-11-25   
9960        9960          9981       23135  218891332  2017-12-15   
9961        9961          9982       23135  221711447  2017-12-27   
9962        9962          9983       23135  242607538  2018-03-12   
9963        9963          9984       23135  245042307  2018-03-20   
9964        9964          9985       23135  247717299  2018-03-29   
9965        9965          9986       23135  251235341  2018-04-07   
9966        9966          9987       23135  256777814  2018-04-23   
9967        9967          9988       23135  259228108  2018-04-30   
9968        9968          9989       23135  262624083  2018-05-09   
9969        9969          9990       23135  264235669  2018-05-13   
9970        9970          9991       23135  266702241  2018-05-20   
9971        9971          9992       23135  269580912  2018-05-27   
9972        9972          9993       23135  277568127  2018-06-16   
9973        9973          9994       23135  281287657  2018-06-24   
9974        9974          9995       23135  292663736  2018-07-18   
9975

In [3]:
listco=dataor1["transcomments"].tolist()
listnon=[]

In [4]:
# -*- coding: utf-8 -*-
import os, requests, uuid, json
subscriptionKey = '...' #USER KEY
base_url = 'https://api.cognitive.microsofttranslator.com'
path = '/translate?api-version=3.0'
params = '&to=en'
constructed_url = base_url + path + params
headers = {
    'Ocp-Apim-Subscription-Key': subscriptionKey,
    'Content-type': 'application/json',
    'X-ClientTraceId': str(uuid.uuid4())
}

In [5]:
for elem in range(len(listco)):
    listco[elem]=str(listco[elem])
    body = [{'text': listco[elem]}]
    request = requests.post(constructed_url, headers=headers, json=body)
    response = request.json()
    if response[0]['detectedLanguage']['language']!="en":
        listnon.append(listco[elem])

In [8]:
print(round(len(listnon)/len(listco)*100,2),"%")

0.71 %


only 0.71% of reviews written in English after translation. 